# HG changeset patch
# User Jaroslav Hajek <highegg@gmail.com>
# Date 1232867096 -3600
#      Sun Jan 25 08:04:56 2009 +0100
# Node ID 79845b1793cf079861840b800a51b1b399b2dd63
# Parent  35656d6ad061c9fca04653b5b3d0a4a3e2cd72b2
optimize cell construction from a cs-list

diff --git a/src/ChangeLog b/src/ChangeLog
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,8 +1,12 @@
+2009-01-24  Jaroslav Hajek  <highegg@gmail.com>
+
+	* pt-cell.cc (tree_cell::rvalue): Optimize the single row case.
+
 2009-01-24  Jaroslav Hajek  <highegg@gmail.com>
 
 	* ov-struct.cc (octave_struct::subsasgn): Reshape cs-list on rhs 
 	according to the index expression.
 	* ov-cell.cc (octave_cell::subsasgn): Dtto.
 
 2008-12-25 Ben Abbott <bpabbott@mac.com>
 
diff --git a/src/pt-cell.cc b/src/pt-cell.cc
--- a/src/pt-cell.cc
+++ b/src/pt-cell.cc
@@ -55,17 +55,20 @@ tree_cell::rvalue (void)
   int i = 0;
 
   for (iterator p = begin (); p != end (); p++)
     {
       tree_argument_list *elt = *p;
 
       octave_value_list row = elt->convert_to_const_vector ();
       
-      if (nc < 0)
+      if (nr == 1)
+        // Optimize the single row case.
+        val = row.cell_value ();
+      else if (nc < 0)
 	{
 	  nc = row.length ();
 
 	  val = Cell (nr, nc);
 	}
       else
 	{
 	  octave_idx_type this_nc = row.length ();
